File: src/services/rich-media.service.ts
import { Injectable } from '@angular/core';

// ─── Rich Media Card Interfaces ───────────────────────────────────────────────

export interface ThreeJsModel {
    id: string;
    name: string;
    description: string;
    threejsId: string;
    severity?: 'green' | 'yellow' | 'red';
    afflictionHighlight?: string;
    particles?: boolean;
}

export interface WikimediaImage {
    title: string;
    url: string;
    thumbUrl: string;
    descriptionUrl: string;
    credit: string;
    license: string;
}

export interface PubmedCitation {
    pmid: string;
    title: string;
    authors: string;
    journal: string;
    year: string;
    abstract: string;
    url: string;
}

export interface PhilImage {
    id: number;
    url: string;
    thumbUrl: string;
    title: string;
    credit: string;
}

export type RichCardKind = 'model-3d' | 'image-gallery' | 'pubmed-refs' | 'phil-image';

export interface RichMediaCard {
    kind: RichCardKind;
    query: string;
    severity?: 'green' | 'yellow' | 'red';
    afflictionHighlight?: string;
    particles?: boolean;
    // Resolved data (populated after fetching)
    models?: ThreeJsModel[];
    images?: WikimediaImage[];
    citations?: PubmedCitation[];
    philImages?: PhilImage[];
    // Loading state
    loading?: boolean;
    error?: string;
}

// ─── Curated Three.js Procedural Registry ────────────────────────────────────
// Rendered locally using procedural shapes in Medical3DViewerComponent

const THREEJS_REGISTRY: Record<string, ThreeJsModel[]> = {
    'spine': [
        {
            id: 'spine',
            name: 'Vertebral Column',
            description: 'Procedural spine segment',
            threejsId: 'spine'
        }
    ],
    'vertebra': [
        {
            id: 'spine',
            name: 'Vertebra Segment',
            description: 'Procedural vertebral segment',
            threejsId: 'spine'
        }
    ],
    'disc herniation': [
        {
            id: 'spine',
            name: 'Spinal Structure',
            description: 'Procedural spine segment',
            threejsId: 'spine'
        }
    ],
    'heart': [
        {
            id: 'heart',
            name: 'Human Heart',
            description: 'Procedural cardiac form',
            threejsId: 'heart'
        }
    ],
    'brain': [
        {
            id: 'brain',
            name: 'Human Brain',
            description: 'Procedural neurological visualization',
            threejsId: 'brain'
        }
    ],
    'lungs': [
        {
            id: 'lungs',
            name: 'Human Lungs',
            description: 'Procedural pulmonary system',
            threejsId: 'lungs'
        }
    ],
    'default': [
        {
            id: 'generic',
            name: 'Organ System Reference',
            description: 'Procedural internal volume reference',
            threejsId: 'generic'
        }
    ]
};

// ─── Curated PHIL Image Registry ─────────────────────────────────────────────
// Public domain CDC images — no API key required

const PHIL_REGISTRY: Record<string, PhilImage[]> = {
    'spine': [
        { id: 9501, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/9501/9501.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/9501/9501_lores.jpg', title: 'Spinal anatomy diagram', credit: 'CDC/PHIL' },
    ],
    'pain': [
        { id: 23258, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/23258/23258.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/23258/23258_lores.jpg', title: 'Chronic pain clinical assessment', credit: 'CDC/PHIL' },
    ],
    'opioid': [
        { id: 22940, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/22940/22940.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/22940/22940_lores.jpg', title: 'Opioid prescribing awareness', credit: 'CDC/PHIL' },
    ],
    'depression': [
        { id: 23095, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/23095/23095.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/23095/23095_lores.jpg', title: 'Mental health and depression', credit: 'CDC/PHIL' },
    ],
    'default': [
        { id: 11162, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/11162/11162.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/11162/11162_lores.jpg', title: 'Clinical care setting', credit: 'CDC/PHIL' },
    ]
};

// ─── Service ──────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class RichMediaService {

    // ─── Procedural Models (local registry) ───────────────────────────────────

    getThreeJsModels(query: string): ThreeJsModel[] {
        const q = query.toLowerCase();
        for (const key of Object.keys(THREEJS_REGISTRY)) {
            if (key === 'default') continue;
            if (q.includes(key)) return THREEJS_REGISTRY[key];
        }
        return THREEJS_REGISTRY['default'];
    }

    // ─── PHIL (curated registry) ─────────────────────────────────────────────

    getPhilImages(query: string): PhilImage[] {
        const q = query.toLowerCase();
        for (const key of Object.keys(PHIL_REGISTRY)) {
            if (key === 'default') continue;
            if (q.includes(key)) return PHIL_REGISTRY[key];
        }
        return PHIL_REGISTRY['default'];
    }

    // ─── Wikimedia Commons ───────────────────────────────────────────────────

    async searchWikimediaImages(query: string, limit = 6): Promise<WikimediaImage[]> {
        const encoded = encodeURIComponent(`${query} anatomy medical`);
        const url = `https://commons.wikimedia.org/w/api.php?action=query&generator=search&gsrnamespace=6&gsrsearch=${encoded}&gsrlimit=${limit}&prop=imageinfo&iiprop=url|descriptionurl|extmetadata&iiurlwidth=400&format=json&origin=*`;

        try {
            const res = await fetch(url);
            const data = await res.json();
            const pages = data?.query?.pages ?? {};

            return Object.values(pages)
                .map((p: any) => {
                    const ii = p.imageinfo?.[0];
                    if (!ii?.url) return null;
                    const meta = ii.extmetadata ?? {};
                    return {
                        title: p.title?.replace('File:', '') ?? '',
                        url: encodeURI(ii.url ?? ''),
                        thumbUrl: encodeURI(ii.thumburl ?? ii.url ?? ''),
                        descriptionUrl: encodeURI(ii.descriptionurl ?? ''),
                        credit: meta.Credit?.value?.replace(/<[^>]+>/g, '') ?? 'Wikimedia Commons',
                        license: meta.LicenseShortName?.value ?? 'See source'
                    } as WikimediaImage;
                })
                .filter((img): img is WikimediaImage => img !== null)
                .filter(img => img.url.match(/\.(jpg|jpeg|png|svg|webp)$/i));
        } catch {
            return [];
        }
    }

    // ─── PubMed ──────────────────────────────────────────────────────────────

    async searchPubmed(query: string, limit = 3): Promise<PubmedCitation[]> {
        const encoded = encodeURIComponent(query);
        const searchUrl = `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=${encoded}&retmax=${limit}&retmode=json&sort=relevance`;

        try {
            const searchRes = await fetch(searchUrl);
            const searchData = await searchRes.json();
            const ids: string[] = searchData?.esearchresult?.idlist ?? [];
            if (ids.length === 0) return [];

            const summaryUrl = `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=${ids.join(',')}&retmode=json`;
            const summaryRes = await fetch(summaryUrl);
            const summaryData = await summaryRes.json();
            const result = summaryData?.result ?? {};

            return ids.map(id => {
                const doc = result[id];
                if (!doc) return null;
                const authors = (doc.authors ?? []).slice(0, 3).map((a: any) => a.name).join(', ');
                return {
                    pmid: id,
                    title: doc.title ?? '',
                    authors: authors + (doc.authors?.length > 3 ? ' et al.' : ''),
                    journal: doc.source ?? '',
                    year: doc.pubdate?.split(' ')[0] ?? '',
                    abstract: doc.title ?? '', // summary doesn't include abstract — use title for preview
                    url: `https://pubmed.ncbi.nlm.nih.gov/${id}/`
                } as PubmedCitation;
            }).filter((c): c is PubmedCitation => c !== null);
        } catch {
            return [];
        }
    }

    // ─── Resolve a card (fetches live data if needed) ─────────────────────────

    async resolveCard(card: RichMediaCard): Promise<RichMediaCard> {
        switch (card.kind) {
            case 'model-3d':
                return { ...card, models: this.getThreeJsModels(card.query), loading: false };
            case 'phil-image':
                return { ...card, philImages: this.getPhilImages(card.query), loading: false };
            case 'image-gallery':
                return { ...card, images: await this.searchWikimediaImages(card.query), loading: false };
            case 'pubmed-refs':
                return { ...card, citations: await this.searchPubmed(card.query), loading: false };
        }
    }
}

File: pocketgull_flutter/lib/services/rich_media_service.dart
/// Rich Media Service — 3D model registry, PHIL images, Wikimedia, PubMed.
///
/// Flutter parity with Angular `rich-media.service.ts` (292 lines).
/// Provides curated model/image registries and network search methods
/// for Wikimedia Commons images and PubMed citations.
library;

import 'dart:convert';
import 'package:http/http.dart' as http;

// ── Types ──────────────────────────────────────────────────────────

class ThreeJsModel {
  final String id;
  final String name;
  final String description;
  final String threejsId;
  const ThreeJsModel({required this.id, required this.name,
      required this.description, required this.threejsId});
}

class WikimediaImage {
  final String title;
  final String url;
  final String thumbUrl;
  final String descriptionUrl;
  final String credit;
  final String license;
  const WikimediaImage({required this.title, required this.url,
      required this.thumbUrl, required this.descriptionUrl,
      required this.credit, required this.license});
}

class PubmedCitation {
  final String pmid;
  final String title;
  final String authors;
  final String journal;
  final String year;
  final String url;
  const PubmedCitation({required this.pmid, required this.title,
      required this.authors, required this.journal,
      required this.year, required this.url});
}

class PhilImage {
  final int id;
  final String url;
  final String thumbUrl;
  final String title;
  final String credit;
  const PhilImage({required this.id, required this.url,
      required this.thumbUrl, required this.title, required this.credit});
}

// ── Curated Registries ─────────────────────────────────────────────

const _threejsRegistry = <String, List<ThreeJsModel>>{
  'spine': [ThreeJsModel(id: 'spine', name: 'Vertebral Column', description: 'Procedural spine segment', threejsId: 'spine')],
  'heart': [ThreeJsModel(id: 'heart', name: 'Human Heart', description: 'Procedural cardiac form', threejsId: 'heart')],
  'brain': [ThreeJsModel(id: 'brain', name: 'Human Brain', description: 'Procedural neurological visualization', threejsId: 'brain')],
  'lungs': [ThreeJsModel(id: 'lungs', name: 'Human Lungs', description: 'Procedural pulmonary system', threejsId: 'lungs')],
  'default': [ThreeJsModel(id: 'generic', name: 'Organ System Reference', description: 'Procedural internal volume reference', threejsId: 'generic')],
};

const _philRegistry = <String, List<PhilImage>>{
  'spine': [PhilImage(id: 9501, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/9501/9501.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/9501/9501_lores.jpg', title: 'Spinal anatomy diagram', credit: 'CDC/PHIL')],
  'pain': [PhilImage(id: 23258, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/23258/23258.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/23258/23258_lores.jpg', title: 'Chronic pain clinical assessment', credit: 'CDC/PHIL')],
  'default': [PhilImage(id: 11162, url: 'https://wwwn.cdc.gov/phil/PHIL_Images/11162/11162.jpg', thumbUrl: 'https://wwwn.cdc.gov/phil/PHIL_Images/11162/11162_lores.jpg', title: 'Clinical care setting', credit: 'CDC/PHIL')],
};

// ── Service ────────────────────────────────────────────────────────

class RichMediaService {
  List<ThreeJsModel> getThreeJsModels(String query) {
    final q = query.toLowerCase();
    for (final key in _threejsRegistry.keys) {
      if (key == 'default') continue;
      if (q.contains(key)) return _threejsRegistry[key]!;
    }
    return _threejsRegistry['default']!;
  }

  List<PhilImage> getPhilImages(String query) {
    final q = query.toLowerCase();
    for (final key in _philRegistry.keys) {
      if (key == 'default') continue;
      if (q.contains(key)) return _philRegistry[key]!;
    }
    return _philRegistry['default']!;
  }

  /// Searches Wikimedia Commons for medical/anatomical images.
  Future<List<WikimediaImage>> searchWikimediaImages(String query, {int limit = 6}) async {
    final simplified = _simplifyQuery(query);
    var results = await _fetchWikimedia(simplified, limit);
    if (results.isNotEmpty) return results;

    final firstWord = simplified.split(' ').first;
    if (firstWord != simplified) {
      results = await _fetchWikimedia(firstWord, limit);
      if (results.isNotEmpty) return results;
    }

    return _fetchWikimedia(query.split(' ').take(3).join(' '), limit);
  }

  /// Searches PubMed for citations.
  Future<List<PubmedCitation>> searchPubmed(String query, {int limit = 3}) async {
    try {
      final encoded = Uri.encodeComponent(query);
      final searchUrl = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=$encoded&retmax=$limit&retmode=json&sort=relevance';
      final searchRes = await http.get(Uri.parse(searchUrl));
      final searchData = jsonDecode(searchRes.body) as Map<String, dynamic>;
      final ids = List<String>.from((searchData['esearchresult']?['idlist'] as List?) ?? []);
      if (ids.isEmpty) return [];

      final summaryUrl = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=${ids.join(',')}&retmode=json';
      final summaryRes = await http.get(Uri.parse(summaryUrl));
      final summaryData = jsonDecode(summaryRes.body) as Map<String, dynamic>;
      final result = summaryData['result'] as Map<String, dynamic>? ?? {};

      return ids.map((id) {
        final doc = result[id] as Map<String, dynamic>?;
        if (doc == null) return null;
        final authors = (doc['authors'] as List? ?? [])
            .take(3).map((a) => (a as Map)['name'] ?? '').join(', ');
        final totalAuthors = (doc['authors'] as List?)?.length ?? 0;
        return PubmedCitation(
          pmid: id,
          title: doc['title'] as String? ?? '',
          authors: '$authors${totalAuthors > 3 ? " et al." : ""}',
          journal: doc['source'] as String? ?? '',
          year: (doc['pubdate'] as String? ?? '').split(' ').first,
          url: 'https://pubmed.ncbi.nlm.nih.gov/$id/',
        );
      }).whereType<PubmedCitation>().toList();
    } catch (_) {
      return [];
    }
  }

  // ── Helpers ──

  String _simplifyQuery(String query) {
    const stop = {'physical', 'examination', 'assessment', 'evaluation',
        'findings', 'clinical', 'medical', 'anatomy', 'and', 'or', 'of',
        'the', 'a', 'in', 'with', 'for', 'to', 'from', 'by', 'at', 'on',
        'overview', 'management', 'treatment', 'approach', 'general',
        'imaging', 'review', 'related', 'relevant', 'associated'};
    final words = query.toLowerCase()
        .replaceAll(RegExp(r'[^a-z\s]'), ' ')
        .split(RegExp(r'\s+'))
        .where((w) => w.length > 2 && !stop.contains(w))
        .take(2)
        .toList();
    return words.isNotEmpty ? words.join(' ') : query.split(' ').take(2).join(' ');
  }

  Future<List<WikimediaImage>> _fetchWikimedia(String searchTerm, int limit) async {
    try {
      final encoded = Uri.encodeComponent('$searchTerm anatomy');
      final url = 'https://commons.wikimedia.org/w/api.php?action=query&generator=search&gsrnamespace=6&gsrsearch=$encoded&gsrlimit=$limit&prop=imageinfo&iiprop=url|descriptionurl|extmetadata&iiurlwidth=400&format=json&origin=*';
      final res = await http.get(Uri.parse(url));
      final data = jsonDecode(res.body) as Map<String, dynamic>;
      final pages = (data['query']?['pages'] as Map<String, dynamic>?) ?? {};

      return pages.values.map((p) {
        final page = p as Map<String, dynamic>;
        final ii = ((page['imageinfo'] as List?)?.firstOrNull as Map<String, dynamic>?);
        if (ii == null || ii['url'] == null) return null;
        final meta = ii['extmetadata'] as Map<String, dynamic>? ?? {};
        var credit = (meta['Credit'] as Map?)?['value'] as String? ?? 'Wikimedia Commons';
        credit = credit.replaceAll(RegExp(r'<[^>]+>'), '');
        return WikimediaImage(
          title: (page['title'] as String? ?? '').replaceFirst('File:', ''),
          url: ii['url'] as String,
          thumbUrl: (ii['thumburl'] ?? ii['url']) as String,
          descriptionUrl: ii['descriptionurl'] as String? ?? '',
          credit: credit,
          license: (meta['LicenseShortName'] as Map?)?['value'] as String? ?? 'See source',
        );
      }).whereType<WikimediaImage>()
          .where((img) => RegExp(r'\.(jpg|jpeg|png|svg|webp)$', caseSensitive: false).hasMatch(img.url))
          .toList();
    } catch (_) {
      return [];
    }
  }
}
